[#11463] Add other teams/students in section as option in feedback paths

tjtanjin opened the pull request:
Fixes #11463  

**Outline of Solution**

- Add dropdown option for other teams/students in section in feedback paths.

Example views are provided below:

This is how the 2 new options look like:

![creator_perspective](https://user-images.githubusercontent.com/43908963/141593373-884f6db0-3273-4b29-93bf-4e389cc1ea8b.png)

For the remaining views, they are based of Francis who is a student within this demo course - notice the section "Tutorial Group 3" and team "Team A" which Francis is in:

![all_students](https://user-images.githubusercontent.com/43908963/141584859-4e267c26-c7aa-4bbb-95c4-db522389c842.png)

In the feedback session where question 2 is set to be only for students of the same section, this is the view Francis sees when giving feedback:

![charlie_perspective_giver_1](https://user-images.githubusercontent.com/43908963/141590204-adc93423-95ac-476e-8a7e-3e8b9b68e137.png)

In the feedback session where question 2 is set to be only for students of the same section, this is the response view Francis sees:

![charlie_perspective_recipient_1](https://user-images.githubusercontent.com/43908963/141586908-efd642ed-9258-41d5-bfad-2e015aa626b5.png)

In the feedback session where question 3 is set to be only for teams of the same section, this is the view Francis sees when giving feedback:

![charlie_perspective_giver_2](https://user-images.githubusercontent.com/43908963/141590789-a63cb062-9c6b-4a25-8b53-668997b7c4c5.png)

In the feedback session where question 3 is set to be only for teams of the same section, this is the response view Francis sees:

![charlie_perspective_recipient_2](https://user-images.githubusercontent.com/43908963/141587463-18d1a08c-c2c3-4191-9488-d4a8e79b2b0c.png)

damithc commented:
@tjtanjin I'm OK with the UI. 
I'll leave others to check the logic. But remember to make it work when the respondent have to pick `up to n recipients` from the same section. e.g., give feedback up to 5 students in the same section.

tjtanjin commented:
> @tjtanjin I'm OK with the UI. I'll leave others to check the logic. But remember to make it work when the respondent have to pick `up to n recipients` from the same section. e.g., give feedback up to 5 students in the same section.

Hi Prof Damith, may I check on how the selection up `up to n recipients` should work? I initially tested this on my work branch and noticed issues using this option. Following which, I checked out to the master branch to test this option but it seems to also be showing unexpected behaviour as well. I am not sure if I may have modified anything locally but this is what I've observed on the master branch:

Question 1 is being used for testing here, I tried changing the maximum number of students but even though the value supposedly saves, a refresh shows otherwise.

https://user-images.githubusercontent.com/43908963/141615322-94e251f6-2e69-4a7b-beb3-289ec2a1d533.mp4

Question 2 is being used for testing here, this time, I cannot even modify the maximum number of students. I can however, still save it but refreshing reverts the choice to unlimited again.

https://user-images.githubusercontent.com/43908963/141615403-7be5cd80-477b-4751-9153-b6166bdd433f.mp4

Strangely, I cannot spot any noticeable difference between the 2 question types. In fact, question 1 was initially having the exact same issue as question 2 and simply suddenly changed its behaviour as I continued testing. I would like to double check if this is an issue I am facing locally and if so, could anyone possibly point me towards the right direction to debugging this? I am under the impression that if I am within the master branch (with ./gradlew generateTypes and npm run build executed), this should be working as intended. Thank you!

damithc commented:
> Question 2 is being used for testing here, this time, I cannot even modify the maximum number of students. I can however, still save it but refreshing reverts the choice to unlimited again.

I couldn't reproduce this on the production server. I've created an account for you in the production server. You can try to test is over there too.

tjtanjin commented:
> > Question 2 is being used for testing here, this time, I cannot even modify the maximum number of students. I can however, still save it but refreshing reverts the choice to unlimited again.
> 
> I couldn't reproduce this on the production server. I've created an account for you in the production server. You can try to test is over there too.

Hi Prof Damith, I have tested on the production server and was unable to reproduce this issue as well. However, as the latest commit on master seemed to be failing some checks, I checked out to the commit before it and the issue was no longer present. It seems likely then that this bug was introduced after the production update but it might be better to seek a second confirmation.

damithc commented:
> Hi Prof Damith, I have tested on the production server and was unable to reproduce this issue as well. However, as the latest commit on master seemed to be failing some checks, I checked out to the commit before it and the issue was no longer present. It seems likely then that this bug was introduced after the production update but it might be better to seek a second confirmation.

Yes, it could be a bug introduced more recently.

wkurniawan07 commented:
Between latest `release` and latest `master` there is only one PR merged (#11423), however based on the nature of the bug reported, it makes a lot of sense if the cause is that PR. @xinyee20 can you spend some time to look into this?

damithc commented:
> Hi Prof Damith, I have tested on the production server and was unable to reproduce this issue as well. However, as the latest commit on master seemed to be failing some checks, I checked out to the commit before it and the issue was no longer present. It seems likely then that this bug was introduced after the production update but it might be better to seek a second confirmation.

@tjtanjin Please report this as a separate bug. As it is already in the master and breaks a core function, we need to fix it soon.

madanalogy commented:
@tjtanjin the question edit bug should have been resolved with https://github.com/TEAMMATES/teammates/pull/11480, will leave this PR as ongoing as it seems that you are still working on addressing some comments. Feel free to switch the tag if I was mistaken or once the PR is ready for review otherwise.

tjtanjin commented:
> @tjtanjin the question edit bug should have been resolved with #11480, will leave this PR as ongoing as it seems that you are still working on addressing some comments. Feel free to switch the tag if I was mistaken or once the PR is ready for review otherwise.

I have merged the master branch and fixed the custom number of recipients. Tested it out and it is working as expected so the PR is ready for review! :)

xinyee20 commented:
> Between latest `release` and latest `master` there is only one PR merged (#11423), however based on the nature of the bug reported, it makes a lot of sense if the cause is that PR. @xinyee20 can you spend some time to look into this?

Hi my apologies for the really late response, I just saw this from my email. Can I check if this problem is resolved? If not I will look into it.

madanalogy commented:
> Hi my apologies for the really late response, I just saw this from my email. Can I check if this problem is resolved? If not I will look into it.

Should have been resolved with https://github.com/TEAMMATES/teammates/pull/11480

xinyee20 commented:
> > Hi my apologies for the really late response, I just saw this from my email. Can I check if this problem is resolved? If not I will look into it.
> 
> Should have been resolved with #11480

@wkurniawan07 thank you for fixing it!

jianhandev commented on src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java:
```diff
@@ -455,6 +503,26 @@ public boolean sessionHasQuestions(String feedbackSessionName, String courseId)
+            List<StudentAttributes> studentsInTeamForSection =
+                    studentsLogic.getStudentsForSection(giverSection, question.getCourseId());
+            teamToTeamInSectionMembersTable = CourseRoster.buildTeamToMembersTable(studentsInTeamForSection);
+            for (Map.Entry<String, List<StudentAttributes>> team : teamToTeamInSectionMembersTable.entrySet()) {
+                if (isInstructorGiver && !instructorGiver.isAllowedForPrivilege(
+                        team.getValue().iterator().next().getSection(),
+                        question.getFeedbackSessionName(),
+                        Const.InstructorPermissions.CAN_SUBMIT_SESSION_IN_SECTIONS)) {
+                    // instructor can only see teams in allowed sections for him/her
+                    continue;
+                }
+                // Ensure student('s team) does not evaluate own team.
+                if (!giverTeam.equals(team.getKey())) {
+                    // recipientEmail doubles as team name in this case.
+                    recipients.put(team.getKey(), team.getKey());
+                }
+            }
```
This part is identical from the previous case, might have room for refactoring. Perhaps let the `TEAMS` case fall through to  the `TEAMS_IN_SAME_SECTION` case and check for `generateOptionsFor == FeedbackParticipantType.TEAMS_IN_SAME_SECTION` / `generateOptionsFor == FeedbackParticipantType.TEAMS`. Similarly for the `STUDENTS_IN_SAME_SECTION` case above.

tjtanjin replied:
I have updated the code as you have suggested, thank you!

jianhandev reviewed (changes requested):
Overall looks good, in some places the use of fallthrough may help to reduce duplication of code.
